hls-plugin-api/src/Ide/Plugin/Error.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE LambdaCase                #-}
+{-# LANGUAGE OverloadedStrings         #-}
+module Ide.Plugin.Error (
+      -- * Plugin Error Handling API
+    PluginError(..),
+    runExceptT,
+    withExceptT,
+    pluginResponseM,
+    hoistExceptT,
+    hoistMaybeT,
+    handleMaybe,
+    handleMaybeM,
+    getNormalizedFilePathE,
+) where
+
+import           Control.Monad.Extra         (maybeM)
+import           Control.Monad.IO.Class      (MonadIO, liftIO)
+import           Control.Monad.Trans.Class   (lift)
+import           Control.Monad.Trans.Except  (ExceptT (..), mapExceptT,
+                                              runExceptT, throwE, withExceptT)
+import           Control.Monad.Trans.Maybe   (MaybeT, mapMaybeT)
+import qualified Data.Text                   as T
+import           Language.LSP.Protocol.Types
+import           Prettyprinter
+
+-- ----------------------------------------------------------------------------
+-- Plugin Error wrapping
+-- ----------------------------------------------------------------------------
+
+pluginResponseM :: Monad m => (t -> m (Either a b)) -> ExceptT t m b -> m (Either a b)
+pluginResponseM handler act =
+    runExceptT act >>= \case
+        Right r  -> pure $ Right r
+        Left err -> handler err
+
+-- See Note [PluginError]
+data PluginError
+  = -- |PluginInternalError should be used if something has gone horribly wrong.
```

Comment:
as in, a bug in our code?

I feel like I'm missing an error for "something went wrong, but just in a way that can sometimes happen and it's okay"?

Reply:
Yes, I feel if the function you need fails if the user's state is invalid, that's where the dependency failed error comes in. If it's not because of the user's invalid state, then arguably it's a bug, and should never happen, (and can be logged with the PluginInternalError).

Reply:
> If it's not because of the user's invalid state, then arguably it's a bug, and should never happen

I feel like this is a pretty strong statement and I don't know that I would want to commit to it. I think I'd feel a bit more comfortable with a generic "error but not a bug" error, especially since I think being more prescriptive here isn't really buying us anything.

It also seems like "is this a bug in our code or not" is the dimension that we actually care about?

Reply:
Theoretically, we could implement PluginError as a type class, and allow plugins to easily generate their own errors. However, I feel we are gaining here by restricting the possible set of errors. As we saw previously, everyone was either just throwing InternalError for everything, or were throwing errors that were incorrect (InvalidReqest for example). If there is a legitimate case in some plugin for a different error, then it can still easily be added here, but I feel it will keep the codebase saner and more coherent to restrict the errors here (And to keep them all in one place). 

However, I did change the wording on the PluginInternalError haddock to be more balanced. I feel if people would like to throw errors more than not, we can also downgrade PluginInternalError to Warning so as not to show a popup upon failure.